File: Java/813.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public double largestSumOfAverages(int[] A, int K) {
        int N = A.length;
        double[][] cache = new double[N + 1][K + 1];
        
        int sum = 0;
        for (int i = 0; i < N; i++) {
            sum += A[i];
            cache[i + 1][1] = (double) sum / (i + 1);
        }

        double ret = helper(A, cache, N, K);
        return ret;
    }
    
    private double helper(int[] A, double[][] cache, int n, int k) {
        if (k > n) return 0;
        if (cache[n][k] > 0) return cache[n][k];
        
        double ret = 0;
        int sum = 0;
        for (int i = n - 1; i >= k - 1; i--) {
            sum += A[i];
            ret = Math.max(ret, helper(A, cache, i, k - 1) + (double) sum / (n - i));
        }
        
        cache[n][k] = ret;
        return ret;
    }
}
__________________________________________________________________________________________________
sample 35500 kb submission
class Solution {
    public double largestSumOfAverages(int[] A, int K) {
        int N = A.length;
        double[] P = new double[N+1];
        for (int i = 0; i < N; ++i)
            P[i+1] = P[i] + A[i];

        double[] dp = new double[N];
        for (int i = 0; i < N; ++i)
            dp[i] = (P[N] - P[i]) / (N - i);

        for (int k = 0; k < K-1; ++k)
            for (int i = 0; i < N; ++i)
                for (int j = i+1; j < N; ++j)
                    dp[i] = Math.max(dp[i], (P[j]-P[i]) / (j-i) + dp[j]);

        return dp[0];
    }
}
__________________________________________________________________________________________________

File: Python3/813.py
__________________________________________________________________________________________________
sample 116 ms submission
import functools, itertools
memoise = functools.lru_cache(maxsize=None)
class Solution:
    def largestSumOfAverages(self, A: List[int], K: int) -> float:
        P = [0, *itertools.accumulate(A)]
        
        @memoise
        def avg(i, j):
            return (P[j] - P[i]) / (j - i)
        
        @memoise
        def dp(n, k):
            if 1 <= k <= n:
                return max(dp(i, k-1) + avg(i, n) for i in range(k-1, n))
            elif k == n == 0:
                return 0
            else:
                return float('-inf')
        #print(dp.cache_info())
        return dp(len(A), K)
__________________________________________________________________________________________________
sample 13584 kb submission
class Solution:
    def largestSumOfAverages(self, A: List[int], K: int) -> float:
        # A = [1,2,3]
        # K = 3
        def avg(i, j):
            return sum(A[i: j + 1]) / (j - i + 1)
        
        dp = [[0 for k in range(K + 1)] for i in range(len(A))]
        
        for i in range(len(A)):
            dp[i][1] = avg(0, i)
        
        for k in range(2, K + 1):
            for i in range(k - 1, len(A)):
                ans = 0
                for j in range(k - 1, i + 1):
                    ans = max(ans, dp[j - 1][k - 1] + avg(j, i))
                dp[i][k] = ans
        print(dp)
        return dp[-1][K]
__________________________________________________________________________________________________
